crates/ra_vfs/src/io.rs

```diff
@@ -1,55 +1,184 @@
-            .into_iter()
-            .map(handle_task)
-            .try_for_each(|it| output_sender.send(it))
-            .unwrap()
-    })
+impl Worker {
+    pub(crate) fn start() -> Worker {
+        let watcher = Arc::new(Mutex::new(None));
+        let watcher_clone = watcher.clone();
+        let (worker, worker_handle) =
+            thread_worker::spawn("vfs", 128, move |input_receiver, output_sender| {
+                input_receiver
+                    .into_iter()
+                    .map(|t| handle_task(t, &watcher_clone))
+                    .try_for_each(|it| output_sender.send(it))
+                    .unwrap()
+            });
+        match Watcher::start(worker.inp.clone()) {
+            Ok(w) => {
+                watcher.lock().replace(w);
+            }
+            Err(e) => log::error!("could not start watcher: {}", e),
+        };
+        Worker {
+            worker,
+            worker_handle,
+            watcher,
+        }
+    }
+
+    pub(crate) fn sender(&self) -> &Sender<Task> {
+        &self.worker.inp
+    }
+
+    pub(crate) fn receiver(&self) -> &Receiver<TaskResult> {
+        &self.worker.out
+    }
+
+    pub(crate) fn shutdown(self) -> thread::Result<()> {
+        if let Some(watcher) = self.watcher.lock().take() {
+            let _ = watcher.shutdown();
+        }
+        let _ = self.worker.shutdown();
+        self.worker_handle.shutdown()
+    }
 }
 
-fn handle_task(task: Task) -> TaskResult {
-    let Task { root, path, filter } = task;
-    log::debug!("loading {} ...", path.as_path().display());
-    let files = load_root(path.as_path(), &*filter);
-    log::debug!("... loaded {}", path.as_path().display());
-    TaskResult { root, files }
+fn watch(
+    watcher: &Arc<Mutex<Option<Watcher>>>,
+    dir: &Path,
+    filter_entry: &RootFilter,
+    emit_for_existing: bool,
+) {
+    let mut watcher = watcher.lock();
+    let watcher = match *watcher {
+        Some(ref mut w) => w,
+        None => {
+            // watcher dropped or couldn't start
+            return;
+        }
+    };
+    watcher.watch_recursive(dir, filter_entry, emit_for_existing)
+}
+
+fn handle_task(task: Task, watcher: &Arc<Mutex<Option<Watcher>>>) -> TaskResult {
+    match task {
+        Task::AddRoot {
+            root,
+            path,
+            root_filter,
+            nested_roots,
+        } => {
+            watch(watcher, &path, root_filter.as_ref(), false);
+            log::debug!("loading {} ...", path.as_path().display());
+            let files = load_root(
+                path.as_path(),
+                root_filter.as_ref(),
+                nested_roots.as_slice(),
+            );
+            log::debug!("... loaded {}", path.as_path().display());
+            TaskResult::AddRoot(AddRootResult { root, files })
+        }
+        Task::HandleChange(change) => {
+            // forward as is because Vfs has to decide if we should load it
+            TaskResult::HandleChange(change)
```

Comment:
Hm, this looks like one elaborate control flow! I wonder if we could simplify this... At least when I originally thought about it, I've thought of `io` as a separate actor which completely encapsulates file system access and gives very "clean" `TaskResult`s outside:

```rust
enum TaskResult {
    BulkLoadRoot { root: VfsRoot, files: Vec<(Path, String)> },
    AddSingleFile { root: VfsRoot, path: Path, text: String},
    ChangeSingleFile { ... },
    RemoveSingleFile { ... },
}
```

That is, `io` handles all loading and watching of roots,  and  vfs itself then decides if it should ignore `ChangeSingleFile` due to memap or not. Watcher becomes and implementation detail of IO.

I think that probably `Worker` will need to know the set of all roots for this to work, so perhaps we should encapsulate `roots: Arena<VfsRoot, Arc<RootFilter>>,` into `Roots` struct and share it via an `Arc` between vfs and Worker?

Obviously, I don't know the specifics of `notify` API very well, so perhaps this wont' work for some reason?

Reply:
Ok, yes, that sounds better and I think it should work, I will try to implement it tomorrow.

Reply:
Should `Roots` struct be in a `RwLock`? At the moment it can be implemented without it but if the vfs will support adding roots after initialization it will be needed.

Reply:
>Should Roots struct be in a RwLock? At the moment it can be implemented without it but if the vfs will support adding roots after initialization it will be needed.

I'd lean towards sharing an immutable snapshot for starters.

Reply:
This is now implemented.

I had to move the watcher to inside the io workers thread:
https://github.com/rust-analyzer/rust-analyzer/blob/1692a11d09cc3d70dcde10cc896c6f7ebb9eb3c3/crates/ra_vfs/src/io.rs#L50-L51
to have access to the output sender and be able to send the `TaskResult`s directly from the watcher.

I've added `excluded_dirs` to `RootFilter` and also
https://github.com/rust-analyzer/rust-analyzer/blob/1692a11d09cc3d70dcde10cc896c6f7ebb9eb3c3/crates/ra_vfs/src/lib.rs#L63
to make using it with `walkdir` cleaner.

The rest is pretty much as you described it, which is a lot cleaner.

One thing I have found is that `WalkDir` by default doesn't follow symlinks which means that `entry.file_type().is_[dir|file]()` and `entry.path().is_[dir|file]()` won't necessarily match. I'm thinking we should probably follow symlinks but i'm not sure.

Reply:
Yeah, this now looks superb! I’ll do a manual smoke test tomorrow and merge!

I think it’s ok not to follow simlnks for now